viewing_party/party.py

```diff
@@ -1,23 +1,143 @@
 # ------------- WAVE 1 --------------------
 
 def create_movie(title, genre, rating):
-    pass
+    if title == None or genre == None or rating == None:
+        return None
+    return {"title": title, "genre": genre, "rating": rating}
 
+def add_to_watched(user_data, movie):
+    user_data["watched"].append(movie)
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, movie_title):
+    for movie in user_data["watchlist"]:
+        if movie["title"] == movie_title:
+            add_to_watched(user_data, movie)
+            user_data["watchlist"].remove(movie)
+    return user_data
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    rating_sum = 0
+    counter = 0
+    try:
+        for movie in user_data["watched"]:
+            rating_sum += movie["rating"]
+            counter += 1
+        return rating_sum / counter
+    except ZeroDivisionError:
+        return 0
 
+def get_most_watched_genre(user_data):
+    # get and store list of all genre values
+    genre_total_list = []
+    for movie in user_data["watched"]:
+        genre_total_list.append(movie["genre"])
+    # create dictionary of most frequent genres
+    genre_frequencies = {}
+    for genre in genre_total_list:
+        try:
+            genre_frequencies[genre] += 1
+        except KeyError: 
+            genre_frequencies[genre] = 1
+    # identify genre with highest frequency count
+    most_popular_genre = None
+    max_frequency = 0
+    for genre, frequency in genre_frequencies.items():
+        if frequency > max_frequency:
+            most_popular_genre = genre
+            max_frequency = frequency
+    return most_popular_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
-        
+def get_unique_watched(user_data):
+    # start with list of all movies user has watched
+    movie_list = list(user_data["watched"])

```

Comment:
It took me a minute to understand why `list` was being called on objects that were already lists, but I realize it's to create a shallow copy such that removing an element doesn't modify the original list(s). This definitely works but there are some dangers in modifying shallow copies. And again we're seeing the use of exceptions where guard clauses would also work better and have the benefit of being more readable to others. Here's an example of a non-destructive approach to generate a new list rather than modify an existing shallow copy:
```python
    friend_movies = [] 

    for watched_item in user_data['friends']:
        for movie in watched_item['watched']:
            if movie not in friend_movies:
                friend_movies.append(movie)
      
    user_unique_movie_titles = []
    for movie in user_data["watched"]:
        if movie not in friend_movies:
            user_unique_movie_titles.append(movie)

    return user_unique_movie_titles
```